File: packages/cli/templates/scripts/mark-checklist-item.js
#!/usr/bin/env bun
'use strict';
/**
 * mark-checklist-item: marks a single `- [ ]` item as `- [x]` in a spec's
 * `## Checklist` section. Idempotent. Cross-shell (Node only — no shell).
 *
 * Usage:
 *   bun .claude/scripts/mark-checklist-item.js --spec <name> --item "<substring>"
 *   bun .claude/scripts/mark-checklist-item.js --spec <name> --line <N>
 *
 * Resolution:
 *   --spec  Looks up `.claude/spec/active/<name>/spec.md`. Accepts the bare
 *           directory name (e.g. `2026-05-08-foo`) or an absolute path to a
 *           spec.md file.
 *   --item  Substring match — case-sensitive — against the text portion of
 *           each `- [ ] <text>` line within the Checklist section.
 *           First match wins. Errors if no match.
 *   --line  1-based line number within the file (alternative to --item, used
 *           when the item text contains characters that are inconvenient on
 *           the shell). The line at that index must be a `- [ ]` checkbox
 *           inside the Checklist section.
 *
 * Exit codes:
 *   0  Item marked, OR item already `[x]` (no-op).
 *   1  Spec not found, no Checklist section, or item not located.
 *   2  Bad invocation (missing required args, conflicting args).
 *
 * Output:
 *   stdout: one line — `marked` | `already-marked` | `error: <reason>`
 *   stderr: only on error / unexpected conditions.
 *
 * @version 1.0.0
 */

const fs = require('fs');
const path = require('path');

function parseArgs(argv) {
  const out = { spec: null, item: null, line: null, cwd: null };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--spec') { out.spec = argv[++i]; continue; }
    if (a === '--item') { out.item = argv[++i]; continue; }
    if (a === '--line') { out.line = parseInt(argv[++i], 10); continue; }
    if (a === '--cwd') { out.cwd = argv[++i]; continue; }
    if (a === '--help' || a === '-h') { out.help = true; continue; }
  }
  return out;
}

function die(code, msg) {
  process.stdout.write(`error: ${msg}\n`);
  process.exit(code);
}

function resolveSpecPath(spec, cwd) {
  if (!spec) return null;
  // Absolute path to spec.md
  if (path.isAbsolute(spec) && spec.endsWith('.md') && fs.existsSync(spec)) {
    return spec;
  }
  // Bare name → .claude/spec/active/<name>/spec.md
  const active = path.join(cwd, '.claude', 'spec', 'active', spec, 'spec.md');
  if (fs.existsSync(active)) return active;
  // Maybe user passed full directory
  const asDir = path.join(spec, 'spec.md');
  if (fs.existsSync(asDir)) return asDir;
  return null;
}

/**
 * Locate the `## Checklist` section. Returns { startIdx, endIdx } where
 * startIdx is the line index AFTER the `## Checklist` header (the first body
 * line of the section) and endIdx is the line index of the next `## ` header
 * (exclusive) or the end of file.
 */
function findChecklistSection(lines) {
  let startIdx = -1;
  for (let i = 0; i < lines.length; i++) {
    if (/^##\s+Checklist\b/.test(lines[i])) { startIdx = i + 1; break; }
  }
  if (startIdx === -1) return null;
  let endIdx = lines.length;
  for (let i = startIdx; i < lines.length; i++) {
    if (/^##\s/.test(lines[i])) { endIdx = i; break; }
  }
  return { startIdx, endIdx };
}

const CHECKBOX_RE = /^(\s*-\s+)\[([ xX])\](\s+)(.*)$/;

function main() {
  const args = parseArgs(process.argv);
  if (args.help) {
    process.stdout.write('Usage: mark-checklist-item.js --spec <name> (--item <text> | --line <N>) [--cwd <dir>]\n');
    process.exit(0);
  }
  if (!args.spec) die(2, '--spec is required');
  if (!args.item && !args.line) die(2, 'either --item or --line is required');
  if (args.item && args.line) die(2, '--item and --line are mutually exclusive');

  const cwd = args.cwd || process.cwd();
  const specPath = resolveSpecPath(args.spec, cwd);
  if (!specPath) die(1, `spec not found: ${args.spec}`);

  let raw;
  try { raw = fs.readFileSync(specPath, 'utf8'); }
  catch (e) { die(1, `cannot read spec: ${e.message}`); }

  const lines = raw.split('\n');
  const section = findChecklistSection(lines);
  if (!section) die(1, 'no `## Checklist` section in spec');

  let targetIdx = -1;

  if (args.line) {
    const idx = args.line - 1;
    if (idx < section.startIdx || idx >= section.endIdx) {
      die(1, `--line ${args.line} is outside the Checklist section (lines ${section.startIdx + 1}-${section.endIdx})`);
    }
    if (!CHECKBOX_RE.test(lines[idx])) {
      die(1, `--line ${args.line} is not a checkbox`);
    }
    targetIdx = idx;
  } else {
    for (let i = section.startIdx; i < section.endIdx; i++) {
      const m = lines[i].match(CHECKBOX_RE);
      if (!m) continue;
      if (m[2] === ' ' && m[4].includes(args.item)) { targetIdx = i; break; }
    }
    if (targetIdx === -1) {
      // Maybe the only match was already [x] — detect for idempotency.
      for (let i = section.startIdx; i < section.endIdx; i++) {
        const m = lines[i].match(CHECKBOX_RE);
        if (!m) continue;
        if ((m[2] === 'x' || m[2] === 'X') && m[4].includes(args.item)) {
          process.stdout.write('already-marked\n');
          process.exit(0);
        }
      }
      die(1, `no `+'`- [ ]`'+` item matching: ${args.item}`);
    }
  }

  const m = lines[targetIdx].match(CHECKBOX_RE);
  if (m[2] === 'x' || m[2] === 'X') {
    process.stdout.write('already-marked\n');
    process.exit(0);
  }
  lines[targetIdx] = `${m[1]}[x]${m[3]}${m[4]}`;

  try { fs.writeFileSync(specPath, lines.join('\n'), 'utf8'); }
  catch (e) { die(1, `cannot write spec: ${e.message}`); }

  process.stdout.write('marked\n');
  process.exit(0);
}

try { main(); }
catch (err) {
  process.stderr.write(`[mark-checklist-item] ${err.stack || err.message}\n`);
  process.exit(1);
}

File: apps/rt/src/run/mark_checklist_item.rs
//! `mustard-rt run mark-checklist-item` — a port of
//! `scripts/mark-checklist-item.js`.
//!
//! Marks a single `- [ ]` item as `- [x]` in a spec's `## Checklist` section.
//! Idempotent. Resolves a bare spec name to
//! `.claude/spec/active/<name>/spec.md`.
//!
//! Output (stdout): one line — `marked` | `already-marked` | `error: <reason>`.
//! Exit codes: 0 success/no-op, 1 not-found/no-section/not-located, 2 bad args.

use std::path::{Path, PathBuf};

/// Print `error: <msg>` and exit with `code`.
fn die(code: i32, msg: &str) -> ! {
    println!("error: {msg}");
    std::process::exit(code);
}

/// Resolve a spec argument to a `spec.md` path.
fn resolve_spec_path(spec: &str, cwd: &Path) -> Option<PathBuf> {
    let p = Path::new(spec);
    if p.is_absolute() && spec.ends_with(".md") && p.exists() {
        return Some(p.to_path_buf());
    }
    let active = cwd
        .join(".claude")
        .join("spec")
        .join("active")
        .join(spec)
        .join("spec.md");
    if active.exists() {
        return Some(active);
    }
    let as_dir = Path::new(spec).join("spec.md");
    if as_dir.exists() {
        return Some(as_dir);
    }
    None
}

/// Locate the `## Checklist` section. Returns `(start_idx, end_idx)` where
/// `start_idx` is the first body line after the header and `end_idx` is the
/// next `## ` header (exclusive) or end-of-file.
fn find_checklist_section(lines: &[&str]) -> Option<(usize, usize)> {
    let start = lines.iter().position(|l| {
        // `^##\s+Checklist\b`
        l.strip_prefix("##")
            .map(|r| {
                let t = r.trim_start_matches([' ', '\t']);
                t.len() != r.len()
                    && {
                        let lower = t.to_lowercase();
                        lower.strip_prefix("checklist").map_or(false, |tail| {
                            tail.chars()
                                .next()
                                .map_or(true, |c| !(c.is_ascii_alphanumeric() || c == '_'))
                        })
                    }
            })
            .unwrap_or(false)
    })? + 1;
    let mut end = lines.len();
    for (i, l) in lines.iter().enumerate().skip(start) {
        // `^##\s`
        if l.strip_prefix("##").map(|r| r.starts_with([' ', '\t'])).unwrap_or(false) {
            end = i;
            break;
        }
    }
    Some((start, end))
}

/// Parsed checkbox line: `(prefix, state, gap, text)`.
struct Checkbox<'a> {
    prefix: &'a str,
    state: char,
    gap: &'a str,
    text: &'a str,
}

/// Parse a `^(\s*-\s+)\[([ xX])\](\s+)(.*)$` checkbox line.
fn parse_checkbox(line: &str) -> Option<Checkbox<'_>> {
    let trimmed_start = line.len() - line.trim_start().len();
    let after_ws = &line[trimmed_start..];
    let rest = after_ws.strip_prefix('-')?;
    if !rest.starts_with([' ', '\t']) {
        return None;
    }
    let dash_gap_len = rest.len() - rest.trim_start_matches([' ', '\t']).len();
    let prefix_end = trimmed_start + 1 + dash_gap_len;
    let body = &line[prefix_end..];
    let inner = body.strip_prefix('[')?;
    let state = inner.chars().next()?;
    if !matches!(state, ' ' | 'x' | 'X') {
        return None;
    }
    let after_state = &inner[state.len_utf8()..];
    let after_bracket = after_state.strip_prefix(']')?;
    if after_bracket.is_empty() || !after_bracket.starts_with([' ', '\t']) {
        return None;
    }
    let gap_len = after_bracket.len() - after_bracket.trim_start_matches([' ', '\t']).len();
    let text = &after_bracket[gap_len..];
    Some(Checkbox {
        prefix: &line[..prefix_end],
        state,
        gap: &after_bracket[..gap_len],
        text,
    })
}

/// Dispatch `mustard-rt run mark-checklist-item`.
pub fn run(spec: Option<&str>, item: Option<&str>, line: Option<usize>, cwd_arg: Option<&str>) {
    let Some(spec) = spec else {
        die(2, "--spec is required");
    };
    if item.is_none() && line.is_none() {
        die(2, "either --item or --line is required");
    }
    if item.is_some() && line.is_some() {
        die(2, "--item and --line are mutually exclusive");
    }

    let cwd = cwd_arg
        .map(PathBuf::from)
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")));
    let Some(spec_path) = resolve_spec_path(spec, &cwd) else {
        die(1, &format!("spec not found: {spec}"));
    };

    let raw = match std::fs::read_to_string(&spec_path) {
        Ok(r) => r,
        Err(e) => die(1, &format!("cannot read spec: {e}")),
    };
    let mut lines: Vec<String> = raw.split('\n').map(String::from).collect();
    let line_refs: Vec<&str> = lines.iter().map(String::as_str).collect();
    let Some((start, end)) = find_checklist_section(&line_refs) else {
        die(1, "no `## Checklist` section in spec");
    };

    let target_idx: usize = if let Some(n) = line {
        let idx = n.wrapping_sub(1);
        if n == 0 || idx < start || idx >= end {
            die(
                1,
                &format!(
                    "--line {n} is outside the Checklist section (lines {}-{end})",
                    start + 1
                ),
            );
        }
        if parse_checkbox(&lines[idx]).is_none() {
            die(1, &format!("--line {n} is not a checkbox"));
        }
        idx
    } else {
        let item = item.unwrap_or("");
        let mut found: Option<usize> = None;
        for i in start..end {
            if let Some(cb) = parse_checkbox(&lines[i]) {
                if cb.state == ' ' && cb.text.contains(item) {
                    found = Some(i);
                    break;
                }
            }
        }
        match found {
            Some(i) => i,
            None => {
                // Idempotency: was the only match already `[x]`?
                for i in start..end {
                    if let Some(cb) = parse_checkbox(&lines[i]) {
                        if (cb.state == 'x' || cb.state == 'X') && cb.text.contains(item) {
                            println!("already-marked");
                            std::process::exit(0);
                        }
                    }
                }
                die(1, &format!("no `- [ ]` item matching: {item}"));
            }
        }
    };

    let new_line = {
        let cb = match parse_checkbox(&lines[target_idx]) {
            Some(c) => c,
            None => die(1, "target line is not a checkbox"),
        };
        if cb.state == 'x' || cb.state == 'X' {
            println!("already-marked");
            std::process::exit(0);
        }
        format!("{}[x]{}{}", cb.prefix, cb.gap, cb.text)
    };
    lines[target_idx] = new_line;

    if let Err(e) = std::fs::write(&spec_path, lines.join("\n")) {
        die(1, &format!("cannot write spec: {e}"));
    }
    println!("marked");
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    fn write_spec(body: &str) -> (tempfile::TempDir, PathBuf) {
        let dir = tempdir().unwrap();
        let path = dir.path().join("spec.md");
        std::fs::write(&path, body).unwrap();
        (dir, path)
    }

    #[test]
    fn parses_checkbox_lines() {
        let cb = parse_checkbox("  - [ ] do the thing").unwrap();
        assert_eq!(cb.state, ' ');
        assert_eq!(cb.text, "do the thing");
        assert!(parse_checkbox("- not a checkbox").is_none());
    }

    #[test]
    fn finds_checklist_section() {
        let lines = vec!["# Spec", "## Checklist", "- [ ] a", "## Next"];
        let (start, end) = find_checklist_section(&lines).unwrap();
        assert_eq!((start, end), (2, 3));
    }

    #[test]
    fn marks_item_by_substring() {
        let (_d, path) = write_spec("## Checklist\n- [ ] alpha\n- [ ] beta\n");
        let mut lines: Vec<String> =
            std::fs::read_to_string(&path).unwrap().split('\n').map(String::from).collect();
        let refs: Vec<&str> = lines.iter().map(String::as_str).collect();
        let (start, end) = find_checklist_section(&refs).unwrap();
        let mut idx = None;
        for i in start..end {
            if let Some(cb) = parse_checkbox(&lines[i]) {
                if cb.state == ' ' && cb.text.contains("beta") {
                    idx = Some(i);
                }
            }
        }
        let i = idx.unwrap();
        let cb = parse_checkbox(&lines[i]).unwrap();
        lines[i] = format!("{}[x]{}{}", cb.prefix, cb.gap, cb.text);
        assert_eq!(lines[i], "- [x] beta");
    }
}
